### ISSUE TITLE
Using messages as map value

### ISSUE THREAD

**Issue by hugebdu** [bufbuild/protobuf-es#465]
There's sort of inconsistency between "simple" wrapper typed field and repeated/map with a wrapper type, for example
given protobuf

```protobuf
message Foo {
  google.protobuf.StringValue one_string = 1;
  repeated google.protobuf.StringValue many_strings = 2;
}
``` 

the generated message type class looks like 

```typescript
class Foo extends Message<Foo> {
  oneString?: string;
  manyStrings: StringValue[];
}
```

one could argue `manyStrings` type should be `Array<string | undefined>` to make it consistent with `oneString` type.

What's the reason for keeping wrapper types for `repeated` and `map`?

Thanks.

**Comment by smaye81** [comment]
Hi @hugebdu. I think the basic reason is that it doesn't make a ton of sense to have `<string | undefined>` in a `repeated` field or in a `map`. The `StringValue` wrapper is just a way of distinguishing between 'this was intentionally set to empty spaces' vs. 'this wasn't set at all'  so in the latter case, you would simply not add it to the collection, right? 

Is this just curiosity or do you have a specific use case that is hampered by this approach?

**Comment by smaye81** [comment]
Yep, I would agree with you there. In fact, it's actually a bigger question regarding how _any_ message should be handled as a map value. The Protobuf site makes a [cursory mention](https://protobuf.dev/programming-guides/proto3/#maps) of how serialization should work and says that it's language-dependent. However, we recognize there are cases where you'd like to set a key with no value and have it behave in a predictable way. Let's keep this issue open while we investigate the best solution for this. Thanks for calling it out.

**Comment by smaye81** [comment]
Hi @hugebdu. We put some careful thought into this internally and after some deliberation, we feel the current behavior of not allowing `undefined` is the right approach.

The tl;dr is that it is not possible to send a meaningful `null/undefined` value on the wire. With a normal repeated message, there is no way in the binary format to encode `null/undefined`. If you try to do this (in any language that allows it), it gets turned into an empty message on the wire and also when it comes back in.

For maps, it would technically be possible to encode an entry with no value, but the Protobuf team probably decided this would be inconsistent with how normal repeated messages work since maps are actually represented as repeated messages. So, basically you can't transmit `null/undefined/nil`, etc. on the wire for any `repeated` message (including maps). It will get changed into an empty message and subsequently when it arrives at its destination.

Because of the above, we feel that not allowing `null/undefined` to be set in the runtime is the most intuitive since it wouldn't lead to surprising behavior when transmitting data. We took a look at other implementations to see whether they allow an unset value in maps in their runtime and it is pretty varied. Some languages allow it, while others do not, so there really isn't any firm precedent to follow.

In summary, the current behavior seems to be the least-surprising based on the wire behavior. I'm going to close this issue since we don't plan to modify this, but I appreciate you calling this out.

**Comment by hugebdu** [comment]
@smaye81 thanks for the update.
